### ISSUE TITLE
Slow when thousands of isolated components

### ISSUE THREAD

**Issue by staltz** [cyclejs/cycle-dom#85]
Reproducible test case: https://jsbin.com/guhizakoxo/edit?html,js,output

In other words, spawn up 1500 instances of a simple component, gather their vtree$ in an array and make that the children of the final vtree. Noticeable lag when dragging a slider in a component.

- - -

I did some experiments, and it doesn't seem to be combineLatest's fault. I think we need to use virtual-dom Thunks to avoid the unnecessary diff of all 1500 children if it is known that only 1 child has changed.

CC @TylorS @Frikki @Keyamoon

**Comment by TylorS** [comment]
@staltz Experiencing the same sort of thing with Motorcycle, lots of lag. I'm definitely going to be looking into solutions.

**Comment by staltz** [comment]
Ok, seems like the diff is not the bottleneck either. For the JSBin above, the bottleneck is frequent reattachment of event listeners using querySelectorAll. @TylorS isn't Motorcycle DOM avoiding these event listener attachments? Why is it slow also with Motorcycle?

**Comment by staltz** [comment]
> I did some experiments, and it doesn't seem to be combineLatest's fault. 

Just to confirm this one: without creating considerable lag to the framerate, RxJS 4 `combineLatest` is able to handle an array of many thousands (~8000) of Observables, where one of them emits every 20ms. So definitely combineLatest is fine for this (and similar) performance problems.

**Comment by TylorS** [comment]
@staltz 40% of the time is spent running `selectorParser()` from `snabbdom-selector`. I think I've made some improvements there by multicast()ing a couple of streams that probably should have been multicasted to begin with. I'm really still investigating.

**Comment by TylorS** [comment]
For Motorcycle-DOM the avoidance of reattaching the event listeners happens at the very end of the stream, right before the creation of an event listener but all the work that happens prior to that still has to occur *n* times where *n* is the number of DOM nodes on every update.

**Comment by TylorS** [comment]
EDIT: Updated my last comment to be correct ^

**Comment by TylorS** [comment]
@staltz After a couple of hours of playing around, I think I know our issue, and it's not really transposition.
48% of the time in the example  @keyamoon gave us is spent on the functions called by `DOM.select()` and roughly the same percentage goes for Motorcycle. This particular example causes DOM.select() to occur 3000 times total, 1500 times to isolate the components, and 1500 times for each component. Every time that you try to move a slider all 3000 calls must be repeated.

Proof:
https://youtu.be/vITp2f9iBwA

The above was created by adding
`.do(x => console.log(x))`
to rootEl$ in makeElementSelector() of cycle-dom.

**Comment by TylorS** [comment]
Okay, so I **might** have found a solution, using event delegation. It only requires changes to `render-dom.js`. There is no need for test modification etc.

Here's a gist of what I came up with https://.github.com/TylorS/4f34c2f8cc6d00c22efe

Moving a slider around in the example @keyamoon gave us, it renders at ~25FPS, which is about where (or better than) Cycle-DOM has always maxed out for me in the performance test.

Sleep now :+1:

**Comment by staltz** [comment]
@TylorS I love your hard work on this one. Thanks!

**Comment by TylorS** [comment]
No problem. It's be fun trying to reach a solution. I think this is pretty close. I'll have to back track a lot in Motorcycle-DOM to support this, it's pretty dependent on querySelectorAll.

Basically the saving grace is, DOM.select() just builds up the namespace, and when events() is called an event listener is attached to the root element, and the events are filtered (hopefully) accordingly based on the namespace and latest selector. The really cool thing though, is this allows a distinctUntilChanged() right before the event listener is attached, because if the selector hasn't changed there is no need to filter any differently.

The same really goes for calling .observable, if its never observed, it shouldn't do any work, and it still works by allowing select() to just build up the namespace.

To sum it all up, basically DOM.select() is made into a super inexpensive operation, and expensive operations are avoided until the last second, and only if necessary.

**Comment by TylorS** [comment]
@staltz I've also got this working for Motorcycle-DOM. Even larger improvements there, generally because the performance was seriously worse with so many event listeners.

Anyways, would you like a PR for this? Or are you thinking that maybe TS + Snabbdom is the next major release of Cycle-DOM ?

**Comment by staltz** [comment]
> Anyways, would you like a PR for this?

No need for a PR, I got your gist and I'm using it as a basis.

Anyway, super thanks for wielding through to the hard bits. For instance the distinctUntilChanged made a lot of sense.

**Comment by TylorS** [comment]
Great, I'll be happy to see what you can improve upon it.

Not a problem, this is a problem we've known about (kind of) for a while, glad to have it almost done.

The `distinctUntilChanged` is pretty awesome, I wish *most* had an operator that worked the same, It's possible just not as simple.

A future improvement might be to somehow instantiate a single event listener per event type, and sort of memoize it, so it can be reused and just filtered accordingly. I'm not sure how this would look though.

Anyways, let me know if I can help out any further.

**Comment by staltz** [comment]
Actually, don't you think the `.distinctUntilChanged(() => selector)` is actually doing the job of simply `.first()`? The `selector` won't actually change (it's not going to be mutated) ever, unless a new call to `events()` happen, but a new call to `events()` is simply going to return a new Observable...

**Comment by staltz** [comment]
I replaced the distinctUntilChanged with `first()` and all tests still passed. Also, it worries me a bit if we can actually trust our tests. It's just a fear, you know, "tests show the presence of bugs, not the absence of bugs".

**Comment by TylorS** [comment]
Yeah you're right, I guess I also `distinctUntilChanged()` might be a bit overkill. Probably solves a problem that probably can't occur.

Yeah, I'm worried about that as well. Maybe a couple of rc's would be the best bet for this feature, it's a pretty radical internal change.

**Comment by staltz** [comment]
Yeah, RC approach can work. I'm mostly worried about moving away from the querySelector approach and moving to an approach that patches VTrees before they reach diff & patch. I liked how snabbdom was different: select vtree, then go to its DOM element, then attach event listener on that.

**Comment by TylorS** [comment]
I'm not sure if that's really an issue? The root element always exists, which is where the event listener is attached. 

Is there something I'm missing from that?

**Comment by staltz** [comment]
I don't know, I'll try to find a test case.

**Comment by TylorS** [comment]
Alright well, I hope it gets us off in the right direction at least. Let me know if there's anything else I can help with.

**Comment by staltz** [comment]
Now that I let the new code sink into my brain, this was just event delegation. It's known to work well. However it depends on event bubbling. Does `useCapture = true` make this method break? I noticed we don't have any tests for the useCapture parameter. @Frikki can you provide a small test case?

**Comment by TylorS** [comment]
Yeah sorry, that's what I called it in the Motorcycle PR. Thought you knew that, my bad. 

In the example @Keyamoon gave me for Motorcycle, which is essentially the same, I did use `useCapture` and it was all in working order. Though I have not used it with the Cycle implementation.

If memory serves me correctly, Frikki wanted `useCapture` to be able to use `blur` events correctly. A test for that could be a good start.

**Comment by staltz** [comment]
Yeah if we get a test for useCapture, then I can add another test and then I think we don't need RC version to release this fix out. :)

**Comment by TylorS** [comment]
I'll work on one getting one for `useCapture`

**Comment by staltz** [comment]
I added this test, just to calm down my nerves:

```js
    it('should catch interaction events from future elements', function (done) {
      function app() {
        return {
          DOM: Rx.Observable.concat(
            Rx.Observable.just(h2('.blesh', 'Blesh')),
            Rx.Observable.just(h3('.blish', 'Blish')).delay(100),
            Rx.Observable.just(h4('.blosh', 'Blosh')).delay(100),
          )
        };
      }
      let {sinks, sources} = Cycle.run(app, {
        DOM: makeDOMDriver(createRenderTarget('parent-002'))
      });
      // Make assertions
      sources.DOM.select('.blosh').events('click').subscribe(ev => {
        assert.strictEqual(ev.type, 'click');
        assert.strictEqual(ev.target.textContent, 'Blosh');
        sources.dispose();
        done();
      });
      sources.DOM.select(':root').observable.skip(3).take(1)
        .subscribe(function (root) {
          let myElement = root.querySelector('.blosh');
          assert.notStrictEqual(myElement, null);
          assert.notStrictEqual(typeof myElement, 'undefined');
          assert.strictEqual(myElement.tagName, 'H4');
          assert.strictEqual(myElement.textContent, 'Blosh');
          assert.doesNotThrow(function () {
            myElement.click();
          });
        });
    });
```

**Comment by TylorS** [comment]
Working test for useCapture
```js
it('should catch a blur event with useCapture', function (done) {
      function app() {
        return {
          DOM: Rx.Observable.just(div('.parent', [
            input('.correct', {type: 'text'}, []),
            input('.wrong', {type: 'text'}, [])
          ]))
        }
      }

      let {sinks, sources} = Cycle.run(app, {
        DOM: makeDOMDriver(createRenderTarget())
      })

      sources.DOM.select('.wrong').events('blur', false).subscribe(assert.fail)

      sources.DOM.select('.correct').events('blur', true).subscribe(ev => {
        assert.strictEqual(ev.type, 'blur')
        assert.strictEqual(ev.target.className, 'correct')
        done()
      })

      sources.DOM.select(':root').observable.skip(1).take(1).subscribe(root => {
        const wrong = root.querySelector('.wrong')
        wrong.focus()
        wrong.blur()
        const correct = root.querySelector('.correct')
        correct.focus()
        correct.blur()
      })
    })
```

**Comment by staltz** [comment]
Done in this commit: https://github.com/cyclejs/cycle-dom/commit/87db2b9e3919dc692bcabe5e58fec04ff5c6229c

I'll make a version later, then close this issue.

**Comment by staltz** [comment]
@TylorS Your help would be appreciated now. It seems the tests are failing consistently in saucelabs for some versions of Firefox. Check this out https://travis-ci.org/cyclejs/cycle-dom/builds/99399492

```
<firefox 39 on Mac 10.10> Rendering :: DOM Driver :: should catch interaction events coming from wrapped View

Error: TypeError: 'matches' called on an object that does not implement interface Element. (https://hbizugkplf.localtunnel.me/__zuul/test-bundle.js:202)
```

Seems to be related to `matchesSelector()`.

**Comment by TylorS** [comment]
@staltz Switch, ev.srcElement to ev.target. ev.srcElement is not a standard apparently.

**Comment by staltz** [comment]
Thanks!

**Comment by TylorS** [comment]
Did that fix it?

**Comment by staltz** [comment]
Yes.

**Comment by TylorS** [comment]
:+1:

**Comment by staltz** [comment]
Resolved in v8.1.0

**Comment by TylorS** [comment]
Nice!
